File: src/platform_context_graph/resolution/shared_projection/partitioning.py
"""Stable partitioning helpers for shared projection domains."""

from __future__ import annotations

import hashlib

from .models import SharedProjectionIntentRow


def partition_for_key(partition_key: str, *, partition_count: int) -> int:
    """Return the stable partition id for one shared projection key."""

    if partition_count <= 0:
        raise ValueError("partition_count must be positive")
    digest = hashlib.sha256(partition_key.encode("utf-8")).digest()
    return int.from_bytes(digest[:8], byteorder="big", signed=False) % partition_count


def rows_for_partition(
    rows: list[SharedProjectionIntentRow],
    *,
    partition_id: int,
    partition_count: int,
) -> list[SharedProjectionIntentRow]:
    """Return intent rows whose partition key belongs to one worker partition."""

    return [
        row
        for row in rows
        if partition_for_key(row.partition_key, partition_count=partition_count)
        == partition_id
    ]

File: go/internal/reducer/partitioning.go
package reducer

import (
	"crypto/sha256"
	"encoding/binary"
	"fmt"
)

// PartitionForKey returns the stable partition id for one shared projection key.
// Uses SHA256 and takes the first 8 bytes as big-endian uint64 mod partitionCount.
func PartitionForKey(partitionKey string, partitionCount int) (int, error) {
	if partitionCount <= 0 {
		return 0, fmt.Errorf("partitionCount must be positive, got %d", partitionCount)
	}

	digest := sha256.Sum256([]byte(partitionKey))
	val := binary.BigEndian.Uint64(digest[:8])

	return int(val % uint64(partitionCount)), nil
}
